trilinos: replace pytrilinos2 variant with python

jmlapre opened the pull request:
There is an ill-named variant "python" that enables the pytrilinos1 variant.  This made it through our testing but broke on our actual CI test machines.

jmlapre committed:
depend_on python

There is an ill-named variant "python" that enables the pytrilinos1
variant.  This made it through our testing but broke on our actual
CI test machines.

psakievich reviewed (approved):
LGTM

sethrj reviewed (commented):
I should've caught this in the first review: the `+python` variant for a package typically means python bindings are generated. Is pytrilinos 1 going to be removed? Maybe the solution is to update the code so that `@14.5:` uses pytrilinos2 when python is enabled.

Are pytrilinos1 and pytrilinos2 compatible?

jmlapre commented:
> I should've caught this in the first review: the `+python` variant for a package typically means python bindings are generated. Is pytrilinos 1 going to be removed? Maybe the solution is to update the code so that `@14.5:` uses pytrilinos2 when python is enabled.
> 
> Are pytrilinos1 and pytrilinos2 compatible?

As far as I know they are not mutually exclusive which is why our initial testing failed to catch this.  As for the future of pytrilinos1, I'll leave that to @sebrowne.

cgcgcg commented:
I think it might be good if one could enable them independently from each other. As far as I know PyTrilinos has been broken for a while, but we don't want to break the installation for someone who wants to use an older Trilinos.

kliegeois commented:
I confirm that PyTrilinos(1) has been broken for a while (more than a year).
I do not have a strong feeling whether or not enabling one should exclude the other one but I like @sebrowne point of view.
At the end of the day, what I would really like is to be able to enable PyTrilinos2 without enabling PyTrilinos(1).

psakievich commented:
> Long-term PyTrilinos will go away and be replaced by PyTrilinos2. Short-term, we may want to keep both to not break any users of PyTrilinos while we spin up full SQA/testing on PyTrilinos2. Pulling in @kliegeois and @cgcgcg to get their input.

> I confirm that PyTrilinos(1) has been broken for a while (more than a year).
I do not have a strong feeling whether or not enabling one should exclude the other one but I like @sebrowne point of view.
At the end of the day, what I would really like is to be able to enable PyTrilinos2 without enabling PyTrilinos(1).

This might be controversial but why not just have one `pytrilinos` variant and conditionally switch the behavior based on the version?  Seems like maintaining two variants for overlapping functionality plus one path broken and no long term plan to maintain it is bloating things a bit.  

Would anyone use `+pytrilinos(1)` off of the develop branch today?

kliegeois commented:
> > Long-term PyTrilinos will go away and be replaced by PyTrilinos2. Short-term, we may want to keep both to not break any users of PyTrilinos while we spin up full SQA/testing on PyTrilinos2. Pulling in @kliegeois and @cgcgcg to get their input.
> 
> > I confirm that PyTrilinos(1) has been broken for a while (more than a year).
> > I do not have a strong feeling whether or not enabling one should exclude the other one but I like @sebrowne point of view.
> > At the end of the day, what I would really like is to be able to enable PyTrilinos2 without enabling PyTrilinos(1).
> 
> This might be controversial but why not just have one `pytrilinos` variant and conditionally switch the behavior based on the version? Seems like maintaining two variants for overlapping functionality plus one path broken and no long term plan to maintain it is bloating things a bit.
> 
> Would anyone use `+pytrilinos(1)` off of the develop branch today?

I think that it is what @sebrowne and @sethrj proposed as well.

sebrowne commented:
I re-read Seth's comment and yes, I agree that is the best path forwards.  So to summarize:

- Leave the existing variant named `python
- As of `15.0:` it will `Trilinos_ENABLE_PyTrilinos2`
- Prior to 15.0 it will `Trilinos_ENABLE_PyTrilinos`

@jmlapre does this make sense?

psakievich commented:
> I think that it is what @sebrowne and @sethrj proposed as well.

🤦‍♂️ Should have read more closely.  Well we are of one accord,so no controversy to be found.

jmlapre commented:
> I re-read Seth's comment and yes, I agree that is the best path forwards. So to summarize:
> 
>     * Leave the existing variant named `python`
> 
>     * As of `15.0:` it will `Trilinos_ENABLE_PyTrilinos2`
> 
>     * Prior to 15.0 it will `Trilinos_ENABLE_PyTrilinos`
> 
> 
> @jmlapre does this make sense?

Yup!

jmlapre committed:
adjust "python" variant based on Trilinos version

For Trilinos <= 14, enable PyTrilinos(1). For later versions
of Trilinos, enable PyTrilinos2.

We still support directly enabling PyTrilinos2 via the "pytrilinos2"
variant.

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -403,6 +404,7 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
 
     depends_on("adios2", when="+adios2")
+    depends_on("binder@1.3:", when="@15: +python", type="build")
     depends_on("binder@1.3:", when="+pytrilinos2", type="build")
```
Duplicate in here now, remove this and all of the others that have when='+pytrilinos2'

sebrowne reviewed (changes requested):
Remove the pytrilinos2 variant and associate code, only want +python (and then it will do The Right Thing based on the version).

jmlapre committed:
remove pytrilinos2 variant

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -404,8 +402,7 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("adios2", when="+adios2")
-    depends_on("binder@1.3:", when="@15: +python", type="build")
-    depends_on("binder@1.3:", when="+pytrilinos2", type="build")
+    depends_on("binder@1.3:", when="+python", type="build")
```
Here I think you still want `@15: +python`

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -425,21 +422,16 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("mpi", when="+mpi")
-    depends_on("mpi", when="@15: +python")
-    depends_on("mpi", when="+pytrilinos2")
+    depends_on("mpi", when="+python")
```
Same as previous

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -425,21 +422,16 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("parmetis", when="+scorec")
     depends_on("py-mpi4py", when="+mpi+python", type=("build", "run"))
-    depends_on("py-mpi4py", when="+pytrilinos2", type=("build", "run"))
+    depends_on("py-mpi4py", when="+python", type=("build", "run"))
```
This one can stay since PyTrilinos + PyTrilinos2 both need this depends_on()

kliegeois replied:
Actually, the line above could be deleted as well:
```
depends_on("py-mpi4py", when="+mpi+python", type=("build", "run"))
```
as we have:
```
depends_on("py-mpi4py", when="+python", type=("build", "run"))
```

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -425,21 +422,16 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
-    depends_on("py-numpy", when="+pytrilinos2", type=("build", "run"))
-    depends_on("py-pybind11", when="@15: +python", type=("build", "link"))
-    depends_on("py-pybind11", when="+pytrilinos2", type=("build", "link"))
+    depends_on("py-pybind11", when="+python", type=("build", "link"))
```
`@15: +python`

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -448,7 +440,7 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("suite-sparse", when="+suite-sparse")
     depends_on("superlu-dist", when="+superlu-dist")
     depends_on("superlu@4.3 +pic", when="+superlu")
-    depends_on("swig", when="@:14 +python")
```
This stays `@14: +python`

sebrowne commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -757,7 +752,7 @@ def define_enable(suffix, value=None):
                 ]
             )
 
-        if "+pytrilinos2" in spec or "@15: +python" in spec:
```
`if "@15: +python" in spec`

jmlapre committed:
correct depends_on constraints
